### PR TITLE
New lint [`four_forward_slashes`]

### LINKED ISSUE #9212: Lint for 4 / (////) to help detect broken doc comments
### What it does

Detects four consecutive forward slash characters.

This is almost always indicative of a broken doc comment.

### Lint Name

four-forward-slashes

### Category

suspicious

### Advantage

- The doc comment appears as expected.

### Drawbacks

Users may have a non-standard style that uses 4 forward slashes semantically.

### Example

```rust
//// The life total of a unit
struct Life(u8);
```

Was likely intended to be:

```rust
/// The life total of a unit
struct Life(u8);
```

### PR DESCRIPTION
**Pull request by Centri3**
Closes #9212

changelog: New lint [`four_forward_slashes`]

### CODE REVIEW

File: tests/ui/four_forward_slashes.rs [hunk 1]
```diff
@@ -0,0 +1,44 @@
+//// first line borked doc comment. doesn't combust!
```
Comment (Centri3) [inline tests/ui/four_forward_slashes.rs:1]: I think we'll need a separate test file for this, as we can't have `#![...]` below items

Comment (xFrednet) [reply tests/ui/four_forward_slashes.rs:1]: Yeah, sounds reasonable :)

**Comment by xFrednet** [comment]
r$ @blyxyas 

(xFrednet has picked a reviewer for you, use r$ to override)

File: clippy_lints/src/four_forward_slashes.rs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+            };
+            let contents = contents.trim();
+            if contents.is_empty() {
+                break;
+            }
+            if contents.starts_with("////") {
+                let bounds = file.line_bounds(line);
+                let span = Span::new(bounds.start, bounds.end, SyntaxContext::root(), None);
+
+                if snippet_opt(cx, span).is_some_and(|s| s.trim().starts_with("////")) {
```
Comment (blyxyas) [inline clippy_lints/src/four_forward_slashes.rs:65]: Is checking for "////" two times necessary?

File: tests/ui/four_forward_slashes.stderr [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+error: comment with 4 forward slashes (`////`). This looks like a doc comment, but it isn't
+  --> $DIR/four_forward_slashes.rs:20:1
+   |
+LL | / //// whoops
+LL | | //// two borked comments!
+   | |_
+   |
+help: make this a doc comment by removing one `/`
+   |
+LL + /// whoops
```
Comment (blyxyas) [inline tests/ui/four_forward_slashes.stderr:47]: Could this be compacted into a single suggestion / emission?

**Review by blyxyas — COMMENTED** [review]
I cannot stop thinking about [`rustc_ast::util::comments::gather_comments`](https://doc.rust-lang.org/beta/nightly-rustc/rustc_ast/util/comments/fn.gather_comments.html), it returns a `Vec<Comment>` and that could be used to check if the comment starts with `//` or `////` (I'm not sure `Comment` also includes the initial `//`).

It would be simpler, but we'd have to do a benchmark to verify speed. Maybe SPEEDTEST can help with that?

File: tests/ui/four_forward_slashes.stderr (hunk 3 above)
Comment (Centri3) [reply tests/ui/four_forward_slashes.stderr:47]: Tried to, it was `span_suggestion_verbose` before. Would prefer it but not sure if it's possible sadly

File: clippy_lints/src/four_forward_slashes.rs (hunk 2 above)
Comment (Centri3) [reply clippy_lints/src/four_forward_slashes.rs:65]: That's the quasi-`is_from_proc_macro` check, so kinda.

**Comment by Centri3** [comment]
`gather_comments` may work, but from what I can see it'd do a lot more work for the same result. Even just allocating an empty `Vec`, then increasing its capacity every time a comment is found, would be slow.

Perhaps we can use the lexer though. I hadn't thought to and that may be nicer

**Comment by Centri3** [comment]
Unfortunately `gather_comments` won't work. It doesn't give the span of the comment, only the column it starts at (`Comment.pos`).

File: clippy_lints/src/four_forward_slashes.rs (hunk 2 above)
Comment (Centri3) [reply clippy_lints/src/four_forward_slashes.rs]: Ok it's unnecessary since I've just realized the prior would already cover that :sweat_smile:

**Review by blyxyas — APPROVED** [review]
LGTM, Thanks! :heart: 
cc @xFrednet

**Comment by xFrednet** [comment]
I'm sadly busy until Tuesday/Wednesday, but then I'll do the last check. :)

**Pull request merged**